File: Brute_Force/Back_Tracking/13023 ABCDE.py
import sys
from collections import deque

N, M = map(int,input().split())
graph = [[] for _ in range(N)]
visited = [0 for _ in range(N)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

def dfs(i, depth):
    visited[i] = 1
    if depth == 4:
        print(1)
        sys.exit()
    for i in graph[i]:
        if not visited[i]:
            dfs(i, depth + 1)
            visited[i] = 0

for i in range(N):
    dfs(i, 0)
    visited[i] = 0
print(0)

File: Brute_Force/Back_Tracking/13023 ABCDE.java
import java.sql.Array;
import java.sql.SQLOutput;
import java.util.*;
import java.io.*;
public class Main {
    static ArrayList<ArrayList<Integer>> graph;
    static int[] visited;
    static int N, M, a, b;

    public static void dfs(int i, int depth){
        visited[i] = 1;
        if (depth == 4) {
            System.out.println(1);
            System.exit(0);
        }

        for(int g : graph.get(i)){
            if (visited[g] != 0)
                continue;
            dfs(g, depth + 1);
            visited[g] = 0;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        visited = new int[N];
        graph = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < N; i++)
            graph.add(new ArrayList<Integer>());
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        for (int i = 0; i < N; i++)
        {
            dfs(i, 0);
            visited[i] = 0;
        }
        System.out.println(0);
    }
}
